Sources/NIOTransportServices/NIOTSConnectionBootstrap.swift

```diff
@@ -261,6 +274,528 @@ public final class NIOTSConnectionBootstrap {
+            },
+            postRegisterTransformation: { $1.makeSucceededFuture($0) }
+        ).get()
+    }
+
+    /// Specify the `unixDomainSocket` path to connect to for the UDS `Channel` that will be established.
+    ///
+    /// - Parameters:
+    ///   - unixDomainSocketPath: The _Unix domain socket_ path to connect to.
+    ///   - channelInitializer: A closure to initialize the channel. The return value of this closure is returned from the `connect`
+    ///   method.
+    /// - Returns: The result of the channel initializer.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func connect<Output: Sendable>(
+        unixDomainSocketPath: String,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Output>
+    ) async throws -> Output {
+        let address = try SocketAddress(unixDomainSocketPath: unixDomainSocketPath)
+        return try await self.connect(
+            to: address,
+            channelInitializer: channelInitializer
+        )
+    }
+
+    /// Specify the `endpoint` to connect to for the TCP `Channel` that will be established.
+    ///
+    /// - Parameters:
+    ///   - endpoint: The endpoint to connect to.
+    ///   - channelInitializer: A closure to initialize the channel. The return value of this closure is returned from the `connect`
+    ///   method.
+    /// - Returns: The result of the channel initializer.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func connect<Output: Sendable>(
+        endpoint: NWEndpoint,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Output>
+    ) async throws -> Output {
+        try await self.connect0(
+            channelInitializer: channelInitializer,
+            registration: { connectionChannel, promise in
+                connectionChannel.register().whenComplete { result in
+                    switch result {
+                    case .success:
+                        connectionChannel.triggerUserOutboundEvent(
+                            NIOTSNetworkEvents.ConnectToNWEndpoint(endpoint: endpoint),
+                            promise: promise
+                        )
+                    case .failure(let error):
+                        promise.fail(error)
+                    }
+                }
+            },
+            postRegisterTransformation: { $1.makeSucceededFuture($0) }
+        ).get()
+    }
+
+    /// Use a pre-existing `NWConnection` to connect a `Channel`.
+    ///
+    /// - Parameters:
+    ///   - connection: The `NWConnection` to wrap.
+    ///   - channelInitializer: A closure to initialize the channel. The return value of this closure is returned from the `connect`
+    ///   method.
+    /// - Returns: The result of the channel initializer.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func withExistingNWConnection<Output: Sendable>(
```

Comment:
Is the `with` naming appropriate here? Seems like the file descriptor case where we're actually taking ownership of it.

Reply:
I think that seems reasonable. @Lukasa are you fine with renaming this to `takingOwnershipOfNWConnection(_:)` and renaming the listener API to `takingOwnershipOfNWListener(_:)`?